packages/@aws-cdk/aws-cloudfront/lib/origin.ts

```diff
@@ -0,0 +1,171 @@
+import { IBucket } from '@aws-cdk/aws-s3';
+import { Construct } from '@aws-cdk/core';
+import { CfnDistribution } from './cloudfront.generated';
+import { OriginProtocolPolicy } from './distribution';
+import { OriginAccessIdentity } from './origin_access_identity';
+
+/**
+ * Properties to be used to create an Origin. Prefer to use one of the Origin.from* factory methods rather than
+ * instantiating an Origin directly from these properties.
+ *
+ * @experimental
+ */
+export interface OriginProps {
+  /**
+   * The domain name of the Amazon S3 bucket or HTTP server origin.
+   */
+  readonly domainName: string;
+}
+
+/**
+ * Represents a distribution origin, that describes the Amazon S3 bucket, HTTP server (for example, a web server),
+ * Amazon MediaStore, or other server from which CloudFront gets your files.
+ *
+ * @experimental
+ */
+export abstract class Origin extends Construct {
+
+  /**
+   * Creates a pre-configured origin for a S3 bucket.
+   * If this bucket has been configured for static website hosting, then `fromWebsiteBucket` should be used instead.
+   *
+   * An Origin Access Identity will be created and granted read access to the bucket.
+   *
+   * @param bucket the bucket to act as an origin.
+   */
+  public static fromBucket(scope: Construct, id: string, bucket: IBucket): Origin {
+    return new S3Origin(scope, id, {
+      domainName: bucket.bucketRegionalDomainName,
+      bucket,
+    });
+  }
+
+  /**
+   * Creates a pre-configured origin for a S3 bucket, where the bucket has been configured for website hosting.
+   *
+   * @param bucket the bucket to act as an origin.
+   */
+  public static fromWebsiteBucket(scope: Construct, id: string, bucket: IBucket): Origin {
+    return new HttpOrigin(scope, id, {
+      domainName: bucket.bucketWebsiteDomainName,
+      protocolPolicy: OriginProtocolPolicy.HTTP_ONLY, // S3 only supports HTTP for website buckets
+    });
+  }
+
+  /**
```

Comment:
Shouldn't this also have `fromHttp...`?

Reply:
Yes, like the `enabled` flag, just a matter of where to draw the line for the sake of this initial PR. I can add it in, or save it for the next PR. My intent was to cover the most common use case of S3-bucket based distributions with one or two common options and then add the other surface area later, both to get faster feedback and make the PR easier to review.